[Datasets] Add documentation for MongoDB data source (#30264)
ray-project/ray: doc/source/data/creating-datasets.rst

    ray_datasets["train"].take(2)
    # [{'text': ''}, {'text': ' = Valkyria Chronicles III = \n'}]

.. _dataset_mongo_db:

------------
From MongoDB
------------

The Dataset can also be created from MongoDB with :py:class:`~ray.data.read_mongo`.
This interacts with MongoDB in a way similar to filesystems, except here you will
need to specify the MongoDB source by its `uri <https://www.mongodb.com/docs/manual/reference/connection-string/>`__,
`database and collection <https://www.mongodb.com/docs/manual/core/databases-and-collections/>`__,
and specify `pipeline <https://www.mongodb.com/docs/manual/core/aggregation-pipeline/>`__ to run against
the collection. The execution results are then used to create a Dataset.

.. code-block:: python

    import ray

    # Read a local MongoDB.
    ds = ray.data.read_mongo(
        uri="mongodb://localhost:27017",
        database="my_db",
        collection="my_collection",
        pipeline=[{"$match": {"col": {"$gte": 0, "$lt": 10}}}, {"$sort": "sort_col"}],
    )

>     # Reading a remote MongoDB is the same.

Reviewer: nit `Reading` -> `Read` to be consistent.

Author: Using 'read' seems a grammar error?

    ds = ray.data.read_mongo(
        uri="mongodb://username:password@mongodb0.example.com:27017/?authSource=admin",
        database="my_db",
        collection="my_collection",
        pipeline=[{"$match": {"col": {"$gte": 0, "$lt": 10}}}, {"$sort": "sort_col"}],
    )

> 

Reviewer: shall we add example for writing MongoDB as well?

Author: I was hesitating since this is creating Dataset, but added - it's a small piece anyway.

    ray_datasets["train"].take(2)
    # [{'text': ''}, {'text': ' = Valkyria Chronicles III = \n'}]

.. _dataset_mongo_db:

------------
From MongoDB
------------

The Dataset can also be created from MongoDB with :py:class:`~ray.data.read_mongo`.

> This interacts with MongoDB in a way similar to filesystems, except here you will

Reviewer 2: ```suggestion
This interacts with MongoDB similar to the external storage APIs, except here you will
```

`database and collection <https://www.mongodb.com/docs/manual/core/databases-and-collections/>`__,
and specify `pipeline <https://www.mongodb.com/docs/manual/core/aggregation-pipeline/>`__ to run against
the collection. The execution results are then used to create a Dataset.

.. code-block:: python

    import ray

    # Read a local MongoDB.
    ds = ray.data.read_mongo(
        uri="mongodb://localhost:27017",
        database="my_db",
        collection="my_collection",
        pipeline=[{"$match": {"col": {"$gte": 0, "$lt": 10}}}, {"$sort": "sort_col"}],
    )

    # Reading a remote MongoDB is the same.
    ds = ray.data.read_mongo(
        uri="mongodb://username:password@mongodb0.example.com:27017/?authSource=admin",
        database="my_db",
        collection="my_collection",
        pipeline=[{"$match": {"col": {"$gte": 0, "$lt": 10}}}, {"$sort": "sort_col"}],
    )

    # Write back to MongoDB.
    ds.write_mongo(
        MongoDatasource(),
        uri="mongodb://username:password@mongodb0.example.com:27017/?authSource=admin",
        database="my_db",
        collection="my_collection",

>     )

Author: This code is not runnable without a hosted db that is accessible like s3. We inline those code in doc.

>     )

Reviewer 2: Ah, I thought that there was a way to tell the CI to not run a particular code block in those files, but if we're currently just inlining such code snippets, that sounds good!
